### PR TITLE
Actually use a Minion worker for Gru

### PR DESCRIPTION
**Pull request by kraih**
While learning the Gru code i noticed that it's actually not too hard to remove all the workarounds and make `gru run` start a normal Minion worker. With all the available features activated, such as processing each job in a separate process (which will free up the memory each job used after it is done). Currently there is still a hardcoded limit since some tasks can't handle parallel execution yet, but theoretically once that is fixed Gru could also run jobs in parallel (which should make it easier to move more work into background processes).

I've introduced a new `OpenQA::WebAPI::Command` namespace so there is no need anymore to hide secondary classes in the plugins.

There were also some fork related issues in `t/14-grutasks.t` (mock IPC) and `t/lib/OpenQA/Test/Database.t` (didn't handle reconnects) that i've resolved too.

Progress: https://progress.opensuse.org/issues/45290.

### CODE REVIEW

**Comment by kraih** [comment]
Work in progress for now since i'm still cleaning some things up and testing more edge cases.

**Comment by kraih** [comment]
The commands now also have real usage messages, so this works:
```
$ ./script/openqa gru
Usage: APPLICATION gru COMMAND [OPTIONS]

Commands:
 list  List Gru jobs and more
 run   Start Gru worker
See 'APPLICATION gru help COMMAND' for more information on a specific
command.

$ ./script/openqa gru list -h
Usage: APPLICATION gru list [OPTIONS] [IDS]

  script/openqa gru list

Options:
  -A, --attempts <number>     Number of times performing this new job will be
                              attempted, defaults to 1
  -a, --args <JSON array>     Arguments for new job or worker remote control
                              command in JSON format
  -b, --broadcast <command>   Broadcast remote control command to one or more
                              workers
  -d, --delay <seconds>       Delay new job for this many seconds
  -e, --enqueue <task>        New job to be enqueued
  -f, --foreground            Retry job in "minion_foreground" queue and
                              perform it right away in the foreground (very
                              useful for debugging)
  -H, --history               Show queue history
  -h, --help                  Show this summary of available options
      --home <path>           Path to home directory of your application,
                              defaults to the value of MOJO_HOME or
                              auto-detection
  -L, --locks                 List active named locks
  -l, --limit <number>        Number of jobs/workers to show when listing
                              them, defaults to 100
  -m, --mode <name>           Operating mode for your application, defaults to
                              the value of MOJO_MODE/PLACK_ENV or
                              "development"
  -o, --offset <number>       Number of jobs/workers to skip when listing
                              them, defaults to 0
  -P, --parent <id>           One or more jobs the new job depends on
  -p, --priority <number>     Priority of new job, defaults to 0
  -q, --queue <name>          Queue to put new job in, defaults to "default",
                              or list only jobs in these queues
  -R, --retry                 Retry job
      --remove                Remove job
  -S, --state <name>          List only jobs in these states
  -s, --stats                 Show queue statistics
  -t, --task <name>           List only jobs for these tasks
  -U, --unlock <name>         Release named lock
  -w, --workers               List workers instead of jobs, or show
                              information for a specific worker
```

**Comment by kraih** [comment]
Ok, i think this pull request is done now.

File: lib/OpenQA/WebAPI/Command/gru.pm [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+=head1 SYNOPSIS
+
+  Usage: APPLICATION gru COMMAND [OPTIONS]
+
+=head1 DESCRIPTION
+
+L<OpenQA::WebAPI::Command::gru> lists available Gru commands.
+
+=cut
```
Comment (Martchus) [inline lib/OpenQA/WebAPI/Command/gru.pm]: You added 15 lines of comments. Yet the most interesting question is unanswered: What role plays this file in the overall architecture?

File: lib/OpenQA/WebAPI/Command/gru/list.pm [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                                or list only jobs in these queues
+    -R, --retry                 Retry job
+        --remove                Remove job
+    -S, --state <name>          List only jobs in these states
+    -s, --stats                 Show queue statistics
+    -t, --task <name>           List only jobs for these tasks
+    -U, --unlock <name>         Release named lock
+    -w, --workers               List workers instead of jobs, or show
+                                information for a specific worker
+
```
Comment (Martchus) [inline lib/OpenQA/WebAPI/Command/gru/list.pm]: Looks like you've just copied the lines above from somewhere else. Wouldn't it be sufficient just to put a link to that place so we don't have to care to keep it in sync?

File: lib/OpenQA/WebAPI/Command/gru/run.pm [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                                         MOJO_MODE/PLACK_ENV or "development"
+    -o, --oneshot                        Perform all currently enqueued jobs and
+                                         then exit
+    -q, --queue <name>                   One or more queues to get jobs from,
+                                         defaults to "default"
+    -R, --repair-interval <seconds>      Repair interval, up to half of this
+                                         value can be subtracted randomly to
+                                         make sure not all workers repair at the
+                                         same time, defaults to 21600 (6 hours)
+
```
Comment (Martchus) [inline lib/OpenQA/WebAPI/Command/gru/run.pm]: same as the previous comment

File: lib/OpenQA/WebAPI/GruJob.pm [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+package OpenQA::WebAPI::GruJob;
+use Mojo::Base 'Minion::Job';
+
+use OpenQA::Utils 'log_error';
+
+sub execute {
+    my $self = shift;
+
+    # TTL
```
Comment (Martchus) [inline lib/OpenQA/WebAPI/GruJob.pm]: This comment serves no purpose. I usually start comments with a verb in imperative (like Git commit messages) to describe what the subsequent section does. A single word is quite meaningless.

File: lib/OpenQA/WebAPI/Plugin/Gru.pm [hunk 5]
```diff
@@ -19,25 +19,21 @@ package OpenQA::WebAPI::Plugin::Gru;
 use Mojo::Base 'Mojolicious::Plugin';
 
 use Minion;
-use Scalar::Util ();
 use DBIx::Class::Timestamps 'now';
 use Mojo::Pg;
 
-has [qw(app dsn)];
+has app => undef, weak => 1;
+has 'dsn';
```
Comment (Martchus) [inline lib/OpenQA/WebAPI/Plugin/Gru.pm]: I know you didn't introduce this line, but I just want to note that I don't like cryptic abbreviations.

File: t/14-grutasks.t [hunk 6]
```diff
@@ -216,15 +229,15 @@ is(
 );
 
 # empty the tracking arrays before next test
```
Comment (Martchus) [inline t/14-grutasks.t]: This comment does not longer match the subsequent code.

File: t/14-grutasks.t [hunk 7]
```diff
@@ -252,8 +265,8 @@ is(
...
```
Comment (Martchus) [inline t/14-grutasks.t]: same here

File: lib/OpenQA/WebAPI/Command/gru/list.pm (hunk 2 above)
Comment (kraih) [reply lib/OpenQA/WebAPI/Command/gru/list.pm]: This is not just documentation, the options get shown when you do `script/openqa gru list -h`.

File: lib/OpenQA/WebAPI/Command/gru.pm (hunk 1 above)
Comment (kraih) [reply lib/OpenQA/WebAPI/Command/gru.pm]: It literally just lists available Gru commands when you do `script/openqa gru`.

File: lib/OpenQA/WebAPI/Command/gru/list.pm (hunk 2 above)
Comment (Martchus) [reply lib/OpenQA/WebAPI/Command/gru/list.pm]: If that can technically not easily be solved without so much redundancy it is ok.

File: lib/OpenQA/WebAPI/Command/gru/run.pm (hunk 3 above)
Comment (kraih) [reply lib/OpenQA/WebAPI/Command/gru/run.pm]: Same as my previous comment (`script/openqa gru run -h`). In this case however with the addition of `--oneshot`.

File: lib/OpenQA/WebAPI/Command/gru.pm (hunk 1 above)
Comment (Martchus) [reply lib/OpenQA/WebAPI/Command/gru.pm]: I found it kind of weird that the comment says `Usage: APPLICATION gru COMMAND [OPTIONS]`. That let's one think it is just a plain command line application. Looking at the actual code it seems that it instead provides some information which is likely consumed by some other module.

File: lib/OpenQA/WebAPI/Plugin/Gru.pm (hunk 5 above)
Comment (kraih) [reply lib/OpenQA/WebAPI/Plugin/Gru.pm]: The abbreviation `dsn` is actually a Perl DBI standard, and most Perl programmers know immediately what it means.

Comment (Martchus) [reply lib/OpenQA/WebAPI/Plugin/Gru.pm]: This project is not targeting 'most Perl programmers' (whatever that means). It is targeting people who want to do QA work and likely even wish that project was written in a different language. Please keep that always in mind when doing something "Perlish". Likely it is not appropriate after all.

Comment (Martchus) [reply lib/OpenQA/WebAPI/Plugin/Gru.pm]: This `dsn` abbreviation seems actually not Perl specific at all so I'm o.k. with it.

**Pull request merged**